### ISSUE TITLE
angr

### ISSUE THREAD

**Issue by armijnhemel** [NixOS/nixpkgs#67413]
**Project description**
angr is a python framework for analyzing binaries. It combines both static and dynamic symbolic ("concolic") analysis, making it applicable to a variety of tasks.

**Metadata**

* homepage URL: https://angr.io/
* source URL: https://github.com/angr
* license: bsd2
* platforms: linux

**Comment by MetaDark** [comment]
I started looking into this, but it's taking some time. There are a lot of dependencies that need to be added in addition to angr.

**Comment by Pamplemousse** [comment]
I have been writing PRs, for packaging missing `angr` requirements into `pythonPackages`.
Some are being accepted, some are still pending; But unfortunately, some may not have their place in this repo: as per  @jonringer's [comment on #68428](https://github.com/NixOS/nixpkgs/pull/68428#pullrequestreview-316668189-body-html) .



---

I started wondering couple of weeks ago if I should not even put all my work from `angr` in a repo separated from `nixpkgs` (something like mozilla/nixpkgs-mozilla).
This would be much less troublesome to maintain, especially considering the poor state of their engineering practices (although they are deploying efforts to improve it).

However, I do not have time, nor the knowledge, to make such a repository; And it is very unlikely that the core `angr` team will do, as they have a very small user base using NixOS, and other engineering priorities.

In the end, I learned about NUR very recently, so that might be a way to go to get the package available quickly, and leave time to make a solution the `angr` team can have more control on.

**Comment by jonringer** [comment]
although angr looks like an active repo/community, unfortunately cooldict didn't really hit a standard bar with me. Another committer might disagree and feel fine with it, I just didn't :/

it should be easy enough to get a pipenv or virtualenv going, and use pip to install it though

**Comment by Pamplemousse** [comment]
@jonringer, I am not sure how to interpret your last statement:
You mean pip install it as part of the building of the `angr` package?

**Comment by jonringer** [comment]
```
$ nix-shell -p python3Packages.virtualenv
$ virtualenv venv
$ source venv/bin/activate
$ pip install angr
```

**Comment by Pamplemousse** [comment]
FYI, `angr` updated its code not to be dependent on `cooldict` anymore: https://github.com/angr/angr/commit/aace252da92c78dcb9eb70e9bda9ceb2fe4aa17e .

**Comment by Pamplemousse** [comment]
I have started [angr/nixpkgs](https://github.com/angr/nixpkgs)!

As per the discussion above, it seems more pertinent to isolate the "home-developed" `angr` related packages into a dedicated repository, made available on NUR.

Packages for which I had a PR waiting in NixOS/nixpkgs, are already migrated.
Expressions for other `angr` components will follow.

As my development environment is a bit messy right now (using my fork, the normal nixpkgs, plus the new NUR repo in local), I will wait to have tidied a bit before sharing my nix expressions as examples in [angr/nixpkgs](https://github.com/angr/nixpkgs).

Feel free to reach out if you have questions :)

**Comment by risicle** [comment]
Keep persevering @Pamplemousse, this looks really neat.

**Comment by Pamplemousse** [comment]
@risicle Thanks for the support!
As discussed earlier, `angr` has found its place in NUR: https://github.com/angr/nixpkgs.

I will happily take feedback over there: I am still quite a beginner at Nix/Nixpkgs, so reviews would be extra helpful.
You are welcome to open issues or submit PRs for comments and proposals :)

And I have finally added `angr` to it! :tada:

**Comment by Pamplemousse** [comment]
@jonringer, @armijnhemel
Should this issue be closed as `angr` is available via NUR?

**Comment by jonringer** [comment]
I'm not familiar with the package, but if it there's broad appeal, then i would say continue the good fight.

Most of my hesitation was around some of the dependencies which were unmaintained for over a year (such as cooldict, which was later removed as a dependency).

**Comment by risicle** [comment]
I would certainly appreciate having `pyvex` available to me in nixpkgs...

**Comment by risicle** [comment]
(same goes for `archinfo`)

**Comment by jonringer** [comment]
@Pamplemousse I would try to upstream it, my major concern is if it only works on python2, which according to pypi, it does not.

**Comment by Pamplemousse** [comment]
@jonringer `angr` drops its support for Python2 on their latest major release: https://angr.io/blog/moving_to_angr_8/ .

Hence, it only works on Python3!

**Comment by jonringer** [comment]
I think if the PR was updated to the latest version, this is still applicable to be added to nixpkgs

**Comment by Pamplemousse** [comment]
**TL;DR**: I am going to stick with contributing to [angr/nixpkgs](https://github.com/angr/nixpkgs), and won't spend time or energy trying to package `angr` (and its dependencies) here.

---

Because it's being "cutting-edge" by nature (developed in a academic research environment), I don't think it is worth trying to put `angr` (nor the dependencies the `angr` team develops - `archinfo`, `claripy`, `cle`, `pyvex`, etc.) in `nixpkgs`.

  * They have now a "fast" release cycle: they started doing a "stable" release every week, but slowed down lately (unsure why, but might be related to the fact that the process is not completely automated - so dependent on the workload of the core contributors);
  * Their definition of "stable" is... disconcerting: they often rely on non-stable features of their dependencies, or custom forks. See angr/cle#255, angr/cle#257 for recent examples of that;
  * They sometimes introduce dependencies that might not be up to `nixpkgs` expectations: a recent example is [`pyxbe`](https://pypi.org/project/pyxbe/) (one maintainer, created recently, so hard to tell if it won't become stale).

Which means that updating the `angr` package would often requires repackaging dependencies, sometimes creating expression for custom forks.
Plus, if one wanted to keep up with their "fast" release cycle, it would imply a decent amount of maintenance here.

Quoting myself from a comment above, I believe NUR is more adapted.:
> This would be much less troublesome to maintain (less load on nixpkgs, lower standards for custom packages), especially considering the poor state of their engineering practices (although they are deploying efforts to improve it).

**Comment by armijnhemel** [comment]
This makes sense :-)

**Comment by fabaff** [comment]
There was already a fair amount of work done. I will continue that effort.

**Comment by jonringer** [comment]
as long as the package additions don't look like massive maintenance burdens and the upstreams are semi-active, I don't mind bringing in what else is needed.

Originally, I just had issues with `cooldict`, and you can see from the commit history https://github.com/zardus/cooldict/commits/master that it's not super active, and has a few other red flags (no releases, commit history is... weird, profanity for the purpose of being "edgy")

**Comment by fabaff** [comment]
`angr` is now available.

**Comment by risicle** [comment]
Well done everyone involved :clap:

**Comment by DrRuhe** [comment]
It is marked as broken :/ 

Is it because of this: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/angr/default.nix#L121 ?

**Comment by fabaff** [comment]
> It is marked as broken :/

Yes, it marked as broken on `master`. `stable` should be fine. `angr` is sensitive to `unicorn` releases. It could work.

**Comment by gleachkr** [comment]
Is the `versionAtLeast` in the `broken` attribute quite right? `versionAtLeast` is true on an [exact match](https://github.com/NixOS/nixpkgs/blob/5873ae414338dbc4aa264cf3d53c4180dc5b592d/lib/strings.nix#L1642), but I think the version in the string there is the one that angr is [pinning unicorn](https://github.com/angr/angr/blob/30ade5de905ac9e4e68dfdfb5b88ad11aa47c99c/pyproject.toml#L2) at, so an exact match is what would make this non-broken.
